tidb-binlog/driver/reader/reader.go

```diff
@@ -145,88 +131,78 @@ func NewReader(cfg *Config) (r *Reader, err error) {
-		zap.Int32("partition", partition),
-		zap.Int64("ts", ts))
-	offsets, err := seeker.Seek(topic, ts, []int32{partition})
+func (r *Reader) getOffsetByTS(ts int64, topic string, partitions []int32) (offset int64, err error) {
+	offsets, err := r.client.SeekOffsetFromTS(ts, topic, partitions)
 	if err != nil {
 		err = errors.Trace(err)
 		return
 	}
-
 	offset = offsets[0]
-
 	return
 }
 
 func (r *Reader) run() {
 	offset := r.cfg.Offset
 	log.Debug("start at", zap.Int64("offset", offset))
+	consumerChan := make(chan *kafka_consumer.KafkaMsg, r.cfg.getMessageBufferSize())
+	done := make(chan struct{})
 
-	consumer, err := sarama.NewConsumerFromClient(r.client)
-	if err != nil {
-		log.Fatal("create kafka consumer failed", zap.Error(err))
-	}
-	defer consumer.Close()
-	topic, partition := r.getTopic()
-	partitionConsumer, err := consumer.ConsumePartition(topic, partition, offset)
-	if err != nil {
-		log.Fatal("create kafka partition consumer failed", zap.Error(err))
-	}
-	defer partitionConsumer.Close()
-
-	// add select to avoid message blocking while reading
-	for {
-		select {
-		case <-r.stop:
-			// clean environment
-			partitionConsumer.Close()
-			close(r.msgs)
-			log.Info("reader stop to run")
-			return
-		case kmsg := <-partitionConsumer.Messages():
-			log.Debug("get kafka message", zap.Int64("offset", kmsg.Offset))
-			binlog := new(pb.Binlog)
-			err := binlog.Unmarshal(kmsg.Value)
-			if err != nil {
-				log.Warn("unmarshal binlog failed", zap.Error(err))
-				continue
-			}
-			if r.cfg.CommitTS > 0 && binlog.CommitTs <= r.cfg.CommitTS {
-				log.Warn("skip binlog CommitTs", zap.Int64("commitTS", binlog.CommitTs))
-				continue
-			}
-
-			msg := &Message{
-				Binlog: binlog,
-				Offset: kmsg.Offset,
-			}
+	go func() {
+		// add select to avoid message blocking while reading
+		for {
 			select {
-			case r.msgs <- msg:
 			case <-r.stop:
-				// In the next iteration, the <-r.stop would match again and prepare to quit
-				continue
+				// clean environment
+				r.client.Close()
+				close(r.msgs)
+				close(done)
+				log.Info("reader stop to run")
+				return
+
+			case kmsg := <-consumerChan:
+				err := kafka_consumer.TransformMSG(r.client.ConsumerType(), kmsg)
+				if err != nil {
+					log.Warn("transform message failed", zap.Error(err))
+					continue
+				}
+				binlog := new(pb.Binlog)
+				err = binlog.Unmarshal(kmsg.Value)
+				if err != nil {
+					log.Warn("unmarshal binlog failed", zap.Error(err))
+					continue
+				}
+
+				if r.cfg.CommitTS > 0 && binlog.CommitTs <= r.cfg.CommitTS {
+					log.Warn("skip binlog CommitTs", zap.Int64("commitTS", binlog.CommitTs))
+					continue
+				}
+
+				msg := &Message{
+					Binlog: binlog,
+					Offset: kmsg.Offset,
+				}
+				select {
+				case r.msgs <- msg:
+				case <-r.stop:
+					// In the next iteration, the <-r.stop would match again and prepare to quit
+					continue
+				}
 			}
 		}
+	}()
+
+	err := r.client.ConsumeFromOffset(offset, consumerChan, done)
```

Comment:
if it returns error here, how to close the goroutine we create above?

Reply:
Add Close function at [last](https://github.com/pingcap/tidb-tools/pull/292/files#diff-e7fa4af2efadbe7e7310025be0e2c5f1R215)

Reply:
I means returning error

Reply:
this error won't return until it [close()](https://github.com/pingcap/tidb-tools/pull/292/files#diff-e7fa4af2efadbe7e7310025be0e2c5f1R206-R213)